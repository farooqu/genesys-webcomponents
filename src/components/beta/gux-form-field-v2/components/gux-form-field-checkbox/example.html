<h1>gux-form-field-checkbox-beta</h1>

<p>
  slot a input[type=checkbox] element into a gux-form-field-checkbox-beta
  element
</p>

<h2>Example</h2>
<form onchange="notify(event)" oninput="notify(event)">
  <fieldset>
    <legend>Food</legend>

    <gux-form-field-checkbox-beta>
      <input slot="input" type="checkbox" name="food-1[]" value="pizza" />
      <label slot="label">Pizza</label>
    </gux-form-field-checkbox-beta>

    <gux-form-field-checkbox-beta>
      <input
        slot="input"
        type="checkbox"
        name="food-1[]"
        value="pasta"
        checked
      />
      <label slot="label">Pasta</label>
    </gux-form-field-checkbox-beta>

    <gux-form-field-checkbox-beta>
      <input slot="input" type="checkbox" name="food-1[]" value="hamburger" />
      <label slot="label">Hamburger</label>
    </gux-form-field-checkbox-beta>

    <gux-form-field-checkbox-beta>
      <input
        slot="input"
        type="checkbox"
        name="food-1[]"
        value="sandwich"
        disabled
      />
      <label slot="label">Sandwich</label>
    </gux-form-field-checkbox-beta>

    <gux-form-field-checkbox-beta>
      <input slot="input" type="checkbox" name="food-1[]" value="sushi" />
      <label slot="label">Sushi</label>
      <span slot="error">Subject to availibility</span>
    </gux-form-field-checkbox-beta>
  </fieldset>
</form>

<gux-button-slot-beta>
  <button
    type="button"
    onclick="(function () {
      document.querySelector(`input[name='food-1[]'][value='hamburger']`).indeterminate = true;
    })()"
  >
    Set Hamburger field to indeterminate
  </button>
</gux-button-slot-beta>

<h2>gux-form-field examples for comparison</h2>
<form onchange="notify(event)" oninput="notify(event)">
  <fieldset>
    <legend>Food</legend>

    <gux-form-field>
      <input slot="input" type="checkbox" name="food-2[]" value="pizza" />
      <label slot="label">Pizza</label>
    </gux-form-field>

    <gux-form-field>
      <input
        slot="input"
        type="checkbox"
        name="food-2[]"
        value="pasta"
        checked
      />
      <label slot="label">Pasta</label>
    </gux-form-field>

    <gux-form-field>
      <input
        slot="input"
        type="checkbox"
        name="food-2[]"
        value="hamburger"
        id="hamburger"
      />
      <label slot="label">Hamburger</label>
    </gux-form-field>

    <gux-form-field>
      <input
        slot="input"
        type="checkbox"
        name="food-2[]"
        value="sandwich"
        disabled
      />
      <label slot="label">Sandwich</label>
    </gux-form-field>

    <gux-form-field>
      <input slot="input" type="checkbox" name="food-2[]" value="sushi" />
      <label slot="label">Sushi</label>
      <span slot="error">Subject to availibility</span>
    </gux-form-field>
  </fieldset>
</form>

<gux-button-slot-beta>
  <button
    type="button"
    onclick="(function () {
      document.querySelector(`input[name='food-2[]'][value='hamburger']`).indeterminate = true;
    })()"
  >
    Set Hamburger field to indeterminate
  </button>
</gux-button-slot-beta>
