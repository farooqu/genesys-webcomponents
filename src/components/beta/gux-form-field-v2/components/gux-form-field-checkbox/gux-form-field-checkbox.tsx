import { Component, Element, h, Host, JSX, State } from '@stencil/core';

import { OnMutation } from '../../../../../utils/decorator/on-mutation';
import { onDisabledChange } from '../../../../../utils/dom/on-attribute-change';
import { preventBrowserValidationStyling } from '../../../../../utils/dom/prevent-browser-validation-styling';
import { trackComponent } from '../../../../../usage-tracking';

import { GuxFormFieldError } from '../../functional-components/gux-form-field-error/gux-form-field-error';

import { hasErrorSlot, validateFormIds } from '../../gux-form-field.servce';

/**
 * @slot input - Required slot for input tag
 * @slot label - Required slot for label tag
 * @slot error - Optional slot for error message
 */
@Component({
  styleUrl: 'gux-form-field-checkbox.less',
  tag: 'gux-form-field-checkbox-beta',
  shadow: true
})
export class GuxFormFieldCheckbox {
  private input: HTMLInputElement;
  private disabledObserver: MutationObserver;

  @Element()
  private root: HTMLElement;

  @State()
  private disabled: boolean;

  @State()
  private hasError: boolean = false;

  @OnMutation({ childList: true, subtree: true })
  onMutation(): void {
    this.hasError = hasErrorSlot(this.root);
  }

  componentWillLoad(): void {
    this.setInput();

    this.hasError = hasErrorSlot(this.root);

    trackComponent(this.root);
  }

  disconnectedCallback(): void {
    this.disabledObserver.disconnect();
  }

  render(): JSX.Element {
    return (
      <Host
        class={{
          'gux-input-error': this.hasError,
          'gux-disabled': this.disabled
        }}
      >
        <div class="gux-form-field-container">
          <div class="gux-input">
            <slot name="input" onSlotchange={() => this.setInput()} />
            <slot name="label" />
          </div>
          <GuxFormFieldError hasError={this.hasError}>
            <slot name="error" />
          </GuxFormFieldError>
        </div>
      </Host>
    ) as JSX.Element;
  }

  private setInput(): void {
    this.input = this.root.querySelector(
      'input[type="checkbox"][slot="input"]'
    );

    preventBrowserValidationStyling(this.input);

    this.disabled = this.input.disabled;

    this.disabledObserver = onDisabledChange(
      this.input,
      (disabled: boolean) => {
        this.disabled = disabled;
      }
    );

    validateFormIds(this.root, this.input);
  }
}
