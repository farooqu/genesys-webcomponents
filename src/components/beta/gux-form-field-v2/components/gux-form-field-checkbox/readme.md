# gux-form-field-checkbox-beta



<!-- Auto Generated Below -->


## Slots

| Slot      | Description                     |
| --------- | ------------------------------- |
| `"error"` | Optional slot for error message |
| `"input"` | Required slot for input tag     |
| `"label"` | Required slot for label tag     |


## Dependencies

### Depends on

- [gux-icon](../../../../stable/gux-icon)

### Graph
```mermaid
graph TD;
  gux-form-field-checkbox-beta --> gux-icon
  style gux-form-field-checkbox-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
