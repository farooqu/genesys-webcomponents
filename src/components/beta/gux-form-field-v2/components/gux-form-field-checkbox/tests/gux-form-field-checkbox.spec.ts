import { newSpecPage } from '@stencil/core/testing';
import MutationObserver from 'mutation-observer';

import { GuxFormFieldCheckbox } from '../gux-form-field-checkbox';

const components = [GuxFormFieldCheckbox];
const language = 'en';

describe('gux-form-field-checkbox-beta', () => {
  beforeEach(async () => {
    (
      global as NodeJS.Global & {
        MutationObserver: any;
      }
    ).MutationObserver = MutationObserver;
  });

  it('should build', async () => {
    const html = `
      <gux-form-field-checkbox-beta>
        <input slot="input" type="checkbox" name="food-1[]" value="pizza"/>
        <label slot="label">Pizza</label>
      </gux-form-field-checkbox-beta>
    `;
    const page = await newSpecPage({ components, html, language });

    expect(page.rootInstance).toBeInstanceOf(GuxFormFieldCheckbox);
  });

  describe('#render', () => {
    [
      `
        <gux-form-field-checkbox-beta>
          <input slot="input" type="checkbox" name="food-1[]" value="pizza"/>
          <label slot="label">Pizza</label>
        </gux-form-field-checkbox-beta>
      `,
      `
        <gux-form-field-checkbox-beta>
          <input slot="input" type="checkbox" name="food-1[]" value="pizza" disabled/>
          <label slot="label">Pizza</label>
        </gux-form-field-checkbox-beta>
      `,
      `
        <gux-form-field-checkbox-beta>
          <input slot="input" type="checkbox" name="food-1[]" value="pizza"/>
          <label slot="label">Pizza</label>
          <span slot="error">This is an error message</span>
        </gux-form-field-checkbox-beta>
      `,
      `
        <gux-form-field-checkbox-beta>
          <input slot="input" type="checkbox" name="food-1[]" value="pizza" checked/>
          <label slot="label">Pizza</label>
        </gux-form-field-checkbox-beta>
      `,
      `
        <gux-form-field-checkbox-beta>
          <input slot="input" type="checkbox" name="food-1[]" value="pizza"/>
          <label slot="label">Pizza</label>
          <span slot="error">Error message</span>
        </gux-form-field-checkbox-beta>
      `
    ].forEach((html, index) => {
      it(`should render component as expected (${index + 1})`, async () => {
        const page = await newSpecPage({ components, html, language });

        expect(page.root).toMatchSnapshot();
      });
    });
  });
});
