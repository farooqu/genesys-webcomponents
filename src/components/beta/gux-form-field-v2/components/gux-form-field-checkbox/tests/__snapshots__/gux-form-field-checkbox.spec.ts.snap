// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-form-field-checkbox-beta #render should render component as expected (1) 1`] = `
<gux-form-field-checkbox-beta>
  <mock:shadow-root>
    <div class="gux-form-field-container">
      <div class="gux-input">
        <slot name="input"></slot>
        <slot name="label"></slot>
      </div>
      <div class="gux-form-field-error">
        <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
        <div class="gux-message">
          <slot name="error"></slot>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <input id="gux-form-field-input-i" name="food-1[]" slot="input" type="checkbox" value="pizza">
  <label for="gux-form-field-input-i" slot="label">
    Pizza
  </label>
</gux-form-field-checkbox-beta>
`;

exports[`gux-form-field-checkbox-beta #render should render component as expected (2) 1`] = `
<gux-form-field-checkbox-beta class="gux-disabled">
  <mock:shadow-root>
    <div class="gux-form-field-container">
      <div class="gux-input">
        <slot name="input"></slot>
        <slot name="label"></slot>
      </div>
      <div class="gux-form-field-error">
        <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
        <div class="gux-message">
          <slot name="error"></slot>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <input disabled="" id="gux-form-field-input-i" name="food-1[]" slot="input" type="checkbox" value="pizza">
  <label for="gux-form-field-input-i" slot="label">
    Pizza
  </label>
</gux-form-field-checkbox-beta>
`;

exports[`gux-form-field-checkbox-beta #render should render component as expected (3) 1`] = `
<gux-form-field-checkbox-beta class="gux-input-error">
  <mock:shadow-root>
    <div class="gux-form-field-container">
      <div class="gux-input">
        <slot name="input"></slot>
        <slot name="label"></slot>
      </div>
      <div class="gux-form-field-error gux-show">
        <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
        <div class="gux-message">
          <slot name="error"></slot>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="gux-form-field-error-i" id="gux-form-field-input-i" name="food-1[]" slot="input" type="checkbox" value="pizza">
  <label for="gux-form-field-input-i" slot="label">
    Pizza
  </label>
  <span id="gux-form-field-error-i" slot="error">
    This is an error message
  </span>
</gux-form-field-checkbox-beta>
`;

exports[`gux-form-field-checkbox-beta #render should render component as expected (4) 1`] = `
<gux-form-field-checkbox-beta>
  <mock:shadow-root>
    <div class="gux-form-field-container">
      <div class="gux-input">
        <slot name="input"></slot>
        <slot name="label"></slot>
      </div>
      <div class="gux-form-field-error">
        <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
        <div class="gux-message">
          <slot name="error"></slot>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <input checked="" id="gux-form-field-input-i" name="food-1[]" slot="input" type="checkbox" value="pizza">
  <label for="gux-form-field-input-i" slot="label">
    Pizza
  </label>
</gux-form-field-checkbox-beta>
`;

exports[`gux-form-field-checkbox-beta #render should render component as expected (5) 1`] = `
<gux-form-field-checkbox-beta class="gux-input-error">
  <mock:shadow-root>
    <div class="gux-form-field-container">
      <div class="gux-input">
        <slot name="input"></slot>
        <slot name="label"></slot>
      </div>
      <div class="gux-form-field-error gux-show">
        <gux-icon decorative="" icon-name="alert-warning-octogon"></gux-icon>
        <div class="gux-message">
          <slot name="error"></slot>
        </div>
      </div>
    </div>
  </mock:shadow-root>
  <input aria-describedby="gux-form-field-error-i" id="gux-form-field-input-i" name="food-1[]" slot="input" type="checkbox" value="pizza">
  <label for="gux-form-field-input-i" slot="label">
    Pizza
  </label>
  <span id="gux-form-field-error-i" slot="error">
    Error message
  </span>
</gux-form-field-checkbox-beta>
`;
